### ISSUE TITLE
geocoder implementation

### ISSUE THREAD

**Issue by thestinger** [GrapheneOS/os_issue_tracker#23]

**Comment by anupritaisno1** [comment]
You might be interested in https://github.com/GlassROM/android_vendor_flossprebuilts/commits/lineage-17.1/UnifiedNlp

You also need this commit https://github.com/GlassROM/android_vendor_lineage/commit/45facdf664c73fcaaded8a28a9155b84d4d7357f

**Comment by renlord** [comment]
@thestinger has made it clear that the project does not want to use Fdroid privileged extensions.

**Comment by anupritaisno1** [comment]
@renlord not sure what you mean by that

**Comment by OlafPettersson** [comment]
Any thoughts on the constaints? Can the implementation make use of an online service?

**Comment by thestinger** [comment]
Yes but in that case it probably shouldn't be enabled by default unless we can host the service and there should still be a toggle for it.

**Comment by RAYs3T** [comment]
This is also an issue with business calendar 2 - Is the implementation for this feature on the roadmap? Are there work-a-rounds available to get this working for apps?

**Comment by patienttruth** [comment]
Is [OpenCage](https://opencagedata.com/) something that could be useful? I came across it while setting up [OwnTracks](https://github.com/owntracks/android).

**Comment by RAYs3T** [comment]
Would that be something that can now be implemented with the new network location feature? Apps like BusinessCalendar2 still can't show the spots on the maps.

**Comment by geevarghesest** [comment]
I am unable to use a helper app in my pixel8 for my motorcycle's digital cluster because it needs geocoder to work in order to set location during setup. https://play.google.com/store/apps/details?id=suzuki.com.suzuki
Hoping this will be releasing soon

**Comment by An4cl3t0** [comment]
Just reporting I have this problem with Home Assistant, no geocoded location sensor enabled on the Play Store version.

**Comment by patienttruth** [comment]
Did it work immediately after restarting or did you have to wait for "optimizing apps" to finish? I'm only getting coordinates, no address.

**Comment by patienttruth** [comment]
@thestinger looking at Nominatim it seems someone can set up their own instance fairly simply. Any consideration (pros and cons?) of allowing a custom provider as well?

**Comment by patienttruth** [comment]
@thestinger that must be what's happening. Home Assistant is definitely recognizing the service as it doesn't even let me access the settings when I disable the provider.

Is there a field that the Google implementation provides that this one does not? Perhaps it is the "name" key that you mentioned.

**Comment by soupslurpr** [comment]
@patienttruth Nominatim supports location names, but it's possible it doesn't have one for the coordinates you are providing.

**Comment by patienttruth** [comment]
@soupslurpr copy that.

I'm probably referring to the wrong geocoder service. By an app using the Google API do you mean directly to Google instead of via Android?

I'm wondering about what a standard OS provides, as ultimately it would be great to have feature parity instead of relying on each app to do the assembly.

Thanks for taking the time to explain it to me. 🙂

**Comment by soupslurpr** [comment]
@patienttruth Yes, I directly requesting it from Google with an API key.

The standard OS API does not provide a way to supply the full address, apps must do it themselves.

**Comment by soupslurpr** [comment]
@patienttruth you could try to debug by making a manual request using a web browser and observing the results.

https://nominatim.openstreetmap.org/reverse?format=geocodejson&lat=37&lon=4&zoom=18

replace 37 with your actual latitude and 4 with your actual longitude

**Comment by patienttruth** [comment]
@soupslurpr well that provided an assembled address under the key `label`. Anything replaced with `xxx` is a correct value.

```
{"type":"FeatureCollection","geocoding":{"version":"0.1.0","attribution":"Data © OpenStreetMap contributors, ODbL 1.0. http://osm.org/copyright","licence":"ODbL","query":""},"features":[{"type":"Feature","properties":{"geocoding":{"place_id":xxx,"osm_type":"way","osm_id":xxx,"osm_key":"place","osm_value":"house","type":"house","accuracy":0,"label":"Full Assembled Address","housenumber":"xxx","postcode":"xxx","street":"xxx","county":"xxx County","state":"xxx","country":"United States","country_code":"us","admin":{"level6":"xxx County","level4":"xxx"}}},"geometry":{"type": "Point","coordinates": [-xxx, xxx]}}]}
```

**Comment by patienttruth** [comment]
@soupslurpr I'm unsure if you're the right person to ask, but is there an easy way for you to make this API call to the geocoder? It looks like I'd have to compile a app to do so. Thought maybe you'd know an easier way. I can try to go that route if that's what it takes.

```
Geocoder geocoder = new Geocoder(context, Locale.getDefault());
geocoder.getFromLocation(47.6062, -122.3321, 6);
```

**Comment by soupslurpr** [comment]
@patienttruth to get the precise results you would have to compile an app, but you can view the result of what the GrapheneOS geocoder should request with those parameters.

https://nominatim.openstreetmap.org/reverse?format=geocodejson&lat=47.6062&lon=-122.3321&zoom=18

**Comment by soupslurpr** [comment]
@patienttruth what I sent is the data before it's converted by the GrapheneOS geocoder into the format the app would receive it in

**Comment by patienttruth** [comment]
@soupslurpr I think I figured it out after a dev on the Home Assistant Companion app provided some insight, and I wouldn't be surprised if the same piece is missing for the Timeclock app too.

The HA Companion app calls `getAddressLine` and expects an assembled address at `address`. Nominatim provides an address at `label`, although I think that sometimes it provides a place name there instead.

https://developer.android.com/reference/android/location/Address#getAddressLine(int)

https://github.com/home-assistant/android/blob/9100fc3ae6c617a924047bbe6aa8c1c69bb63d64/app/src/main/kotlin/io/homeassistant/companion/android/sensors/GeocodeSensorManager.kt#L137


The Timeclock app calls

`Geocoder geocoder = new Geocoder(context, Locale.getDefault());
geocoder.getFromLocation(47.6062, -122.3321, 6);`

According to ChatGPT the following is an example of what the response would be. I think it's that `AddressLine[0]` that's missing.


```
[
  Address[
    AddressLine[0] = "Seattle, WA 98104, USA",
    AddressLine[1] = "Seattle",
    AddressLine[2] = "King County",
    AddressLine[3] = "Washington",
    AddressLine[4] = "USA",
    CountryName = "United States",
    CountryCode = "US",
    AdminArea = "Washington",
    SubAdminArea = "King County",
    Locality = "Seattle",
    PostalCode = "98104",
    Latitude = 47.6062,
    Longitude = -122.3321
  ],
  Address[
    AddressLine[0] = "Pioneer Square, Seattle, WA, USA",
    Locality = "Seattle",
    AdminArea = "Washington",
    CountryName = "United States",
    CountryCode = "US"
  ],
  Address[
    AddressLine[0] = "Downtown Seattle, Seattle, WA, USA",
    Locality = "Seattle",
    AdminArea = "Washington",
    CountryName = "United States",
    CountryCode = "US"
  ],
  Address[
    AddressLine[0] = "King County, Washington, USA",
    SubAdminArea = "King County",
    AdminArea = "Washington",
    CountryName = "United States",
    CountryCode = "US"
  ],
  Address[
    AddressLine[0] = "Washington, USA",
    AdminArea = "Washington",
    CountryName = "United States",
    CountryCode = "US"
  ],
  Address[
    AddressLine[0] = "United States",
    CountryName = "United States",
    CountryCode = "US"
  ]
]
```

Here is an example that the dev of the Timeclock app provided. Seems pretty similar to the example by ChatGPT.

![ResizedImage_2025-08-20_17-07-49_4144.png](https://github.com/user-attachments/assets/e3c66e05-9677-48ce-8eb8-cbb36ac17056)

His app then parses that list and provides the radio select.


Is there a possibility that the current geocoder doesn't respond to these API calls, and if Nominatim doesn't support this directly, would it be possible to emulate it?

**Comment by soupslurpr** [comment]
@patienttruth I looked into what you suspected and it seems you're correct! Working on that right now!

**Comment by soupslurpr** [comment]
@patienttruth that's great to hear.

That app is likely requesting multiple addresses around a specific coordinate, but Nominatim only supports returning one at this time.

**Comment by patienttruth** [comment]
@soupslurpr I was wondering if that might be the case. At first attempt accuracy was at 1200m, so it should have had a few addresses in the area as properties where I tested are ~50-100m across. I'll give it a couple tries when I'm in a denser area. Thanks for all the work on moving this forward!
